File: sections/67_persistence_rollback_time/problems/b_versioned_range_add/solution.cpp
#include <bits/stdc++.h>
using namespace std;
using int64 = long long;
struct Node {
  int left = 0, right = 0;
  int64 sum = 0, lazy = 0;
};
vector<Node> tree(1);
int build(const vector<int64>& a, int lo, int hi) {
  int node = tree.size();
  tree.push_back({});
  if (lo == hi)
    tree[node].sum = a[lo];
  else {
    int mid = (lo + hi) / 2;
    tree[node].left = build(a, lo, mid);
    tree[node].right = build(a, mid + 1, hi);
    tree[node].sum = tree[tree[node].left].sum + tree[tree[node].right].sum;
  }
  return node;
}
int update(int old, int lo, int hi, int ql, int qr, int64 value) {
  int node = tree.size();
  tree.push_back(tree[old]);
  if (ql <= lo && hi <= qr) {
    tree[node].sum += value * (hi - lo + 1);
    tree[node].lazy += value;
    return node;
  }
  int mid = (lo + hi) / 2;
  if (ql <= mid) tree[node].left = update(tree[old].left, lo, mid, ql, qr, value);
  if (qr > mid) tree[node].right = update(tree[old].right, mid + 1, hi, ql, qr, value);
  tree[node].sum = tree[tree[node].left].sum + tree[tree[node].right].sum +
                   tree[node].lazy * (hi - lo + 1);
  return node;
}
int64 query(int node, int lo, int hi, int ql, int qr, int64 carry = 0) {
  if (ql <= lo && hi <= qr) return tree[node].sum + carry * (hi - lo + 1);
  carry += tree[node].lazy;
  int mid = (lo + hi) / 2, res = 0;
  if (ql <= mid) res += query(tree[node].left, lo, mid, ql, qr, carry);
  if (qr > mid) res += query(tree[node].right, mid + 1, hi, ql, qr, carry);
  return res;
}
int main() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  int n, q;
  if (!(cin >> n >> q)) return 0;
  vector<int64> a(n + 1);
  for (int i = 1; i <= n; i++) cin >> a[i];
  vector<int> roots{build(a, 1, n)};
  while (q--) {
    char type;
    int v, l, r;
    cin >> type >> v >> l >> r;
    if (type == 'A') {
      int64 x;
      cin >> x;
      roots.push_back(update(roots[v], 1, n, l, r, x));
    } else
      cout << query(roots[v], 1, n, l, r) << '\n';
  }
}

File: sections/67_persistence_rollback_time/problems/b_versioned_range_add/solution.py
import sys


def main() -> None:
    data = sys.stdin.buffer.read().split()
    if not data:
        return
    n, q = map(int, data[:2])
    idx = 2
    a = [0] + list(map(int, data[idx : idx + n]))
    idx += n
    left = [0]
    right = [0]
    total = [0]
    lazy = [0]

    def new(old=0):
        left.append(left[old])
        right.append(right[old])
        total.append(total[old])
        lazy.append(lazy[old])
        return len(total) - 1

    def build(lo, hi):
        node = new()
        if lo == hi:
            total[node] = a[lo]
        else:
            mid = (lo + hi) // 2
            left[node] = build(lo, mid)
            right[node] = build(mid + 1, hi)
            total[node] = total[left[node]] + total[right[node]]
        return node

    def update(old, lo, hi, ql, qr, value):
        node = new(old)
        if ql <= lo and hi <= qr:
            total[node] += value * (hi - lo + 1)
            lazy[node] += value
            return node
        mid = (lo + hi) // 2
        if ql <= mid:
            left[node] = update(left[old], lo, mid, ql, qr, value)
        if qr > mid:
            right[node] = update(right[old], mid + 1, hi, ql, qr, value)
        total[node] = (
            total[left[node]] + total[right[node]] + lazy[node] * (hi - lo + 1)
        )
        return node

    def query(node, lo, hi, ql, qr, carry=0):
        if ql <= lo and hi <= qr:
            return total[node] + carry * (hi - lo + 1)
        carry += lazy[node]
        mid = (lo + hi) // 2
        answer = 0
        if ql <= mid:
            answer += query(left[node], lo, mid, ql, qr, carry)
        if qr > mid:
            answer += query(right[node], mid + 1, hi, ql, qr, carry)
        return answer

    roots = [build(1, n)]
    out = []
    for _ in range(q):
        kind = data[idx].decode()
        v, l, r = map(int, data[idx + 1 : idx + 4])
        idx += 4
        if kind == "A":
            value = int(data[idx])
            idx += 1
            roots.append(update(roots[v], 1, n, l, r, value))
        else:
            out.append(str(query(roots[v], 1, n, l, r)))
    print("\n".join(out))


if __name__ == "__main__":
    main()
